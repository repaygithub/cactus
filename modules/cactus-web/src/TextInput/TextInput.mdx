---
name: TextInput
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import TextInput from './TextInput'
import cactusTheme from '@repay/cactus-theme'

# TextInput

### How it looks

<TextInput />

## Basic usage

```jsx
import { TextInput } from '@repay/cactus-web'
...
// Basic usage
<TextInput />

// Success status
<TextInput status="success" />

// Invalid status
<TextInput status="warning" />

// Error status
<TextInput status="error" />

// Disabled
<TextIput disabled={true} />

// Setting placeholder
<TextInput placeholder="Winter is Coming" />
```

## Properties

<PropsTable of={TextInput} />
