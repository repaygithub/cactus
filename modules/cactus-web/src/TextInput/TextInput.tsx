import React from 'react'
import styled, { FlattenInterpolation, ThemeProps, css } from 'styled-components'
import { CactusTheme } from '@repay/cactus-theme'
import { StatusCheck, NotificationAlert, NotificationError } from '@repay/cactus-icons'

export type Status = 'success' | 'invalid' | 'error'

interface TextInputProps
  extends Omit<
    React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>,
    'ref'
  > {
  disabled?: boolean
  status?: Status | null
}

interface InputProps {
  disabled?: boolean
  status?: Status | null
}

interface StyledCheckProps {
  disabled?: boolean
  status?: Status | null
}

interface StyledAlertProps {
  disabled?: boolean
  status?: Status | null
}

interface StyledErrorProps {
  disabled?: boolean
  status?: Status | null
}

type StatusMap = { [K in Status]: FlattenInterpolation<ThemeProps<CactusTheme>> }

const statusMap: StatusMap = {
  success: css`
    border-color: ${p => p.theme.colors.success};
  `,
  invalid: css`
    border-color: ${p => p.theme.colors.warning};
  `,
  error: css`
    border-color: ${p => p.theme.colors.error};
  `,
}

const displayStatus = (props: TextInputProps) => {
  if (props.status && !props.disabled) {
    return statusMap[props.status]
  }
}

const StyledCheck = styled(StatusCheck)<StyledCheckProps>`
  width: 20px;
  height: 20px;
  position: absolute;
  right: 8px;
  color: ${p => p.theme.colors.success};
`

const StyledAlert = styled(NotificationAlert)<StyledAlertProps>`
  width: 24px;
  height: 24px;
  position: absolute;
  right: 8px;
  color: ${p => p.theme.colors.warning};
`

const StyledError = styled(NotificationError)<StyledErrorProps>`
  width: 20px;
  height: 20px;
  position: absolute;
  right: 8px;
  color: ${p => p.theme.colors.error};
`

const TextInputContainer = styled.div`
  position: relative;
  text-align: center;
  display: flex;
  flex-direction: row;
  align-items: center;
`

const Input = styled.input<InputProps>`
  border: 1px solid ${p => (p.disabled ? p.theme.colors.lightGray : p.theme.colors.darkContrast)};
  border-radius: 20px;
  height: 32px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  width: ${p => p.width || 'auto'};
  background-color: ${p => (p.disabled ? p.theme.colors.lightGray : p.theme.colors.white)};

  &:focus {
    border-color: ${p => p.theme.colors.callToAction};
  }

  &::placeholder {
    color: ${p => p.theme.colors.mediumContrast};
    font-style: oblique;
  }

  ${displayStatus}
`

const TextInput = (props: TextInputProps) => {
  const { disabled, status } = props
  return (
    <TextInputContainer>
      <Input {...props} />
      {status === 'success' && !disabled && <StyledCheck disabled={disabled} status={status} />}
      {status === 'invalid' && !disabled && <StyledAlert disabled={disabled} status={status} />}
      {status === 'error' && !disabled && <StyledError disabled={disabled} status={status} />}
    </TextInputContainer>
  )
}

export default TextInput
