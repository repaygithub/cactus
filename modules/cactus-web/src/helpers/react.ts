import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
