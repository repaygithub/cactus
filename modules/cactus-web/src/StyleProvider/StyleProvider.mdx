---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
