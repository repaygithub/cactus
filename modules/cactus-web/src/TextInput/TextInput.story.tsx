import { boolean, select, text } from '@storybook/addon-knobs'
import { Meta } from '@storybook/react/types-6-0'
import React from 'react'

import actions from '../helpers/storybookActionsWorkaround'
import { Status } from '../StatusMessage/StatusMessage'
import TextInput from './TextInput'

export default {
  title: 'TextInput',
  component: TextInput,
} as Meta

const eventLoggers = actions('onChange', 'onFocus', 'onBlur')

type StatusOptions = { [k in Status | 'none']: Status | null }

const statusOptions: StatusOptions = {
  none: null,
  success: 'success',
  warning: 'warning',
  error: 'error',
}

export const BasicUsage = (): React.ReactElement => (
  <TextInput
    disabled={boolean('disabled', false)}
    placeholder={text('placeholder', 'Placeholder')}
    status={select('status', statusOptions, statusOptions.none)}
    {...eventLoggers}
  />
)
