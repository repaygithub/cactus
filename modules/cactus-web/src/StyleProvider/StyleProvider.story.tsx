import React from 'react'

import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import cactusTheme from '@repay/cactus-theme'
import Grid from '../Grid/Grid'
import styled from 'styled-components'
import StyleProvider from './StyleProvider'

const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const BreakpointBox = styled(Box)`
  max-width: 320px;
  width: 50px;
  height: 50px;
  border-radius: 15%

  ${p => p.theme.mediaQueries.small} {
    max-width: 708px;
    width: 75px;
    height 75px;
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    max-width: 964px;
    width: 100px;
    height: 100px;
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    max-width: 1140px;
    width: 150px;
    height: 150px;
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extralarge} {
    max-width: 1380px;
    width: 200px;
    height: 200px;
    border-radius: 50%;
  }
`

storiesOf('Style Provider', module).add('Component Adjusments based on Media Queries', () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <Grid justify="center">
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
      </Grid>
    </StyleProvider>
  )
})
