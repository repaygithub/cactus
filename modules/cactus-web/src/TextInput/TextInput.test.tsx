import { generateTheme } from '@repay/cactus-theme'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import TextInput from './TextInput'

describe('component: TextInput', (): void => {
  test('should render an input with a placeholder', (): void => {
    const { getByPlaceholderText } = render(
      <StyleProvider>
        <TextInput placeholder="hold my place" />
      </StyleProvider>
    )

    expect(getByPlaceholderText('hold my place')).toBeInTheDocument()
  })

  test('should support margin space props', (): void => {
    const { getByPlaceholderText } = render(
      <StyleProvider>
        <TextInput marginTop={4} placeholder="Do I wanna know?" />
      </StyleProvider>
    )

    const textInput = getByPlaceholderText('Do I wanna know?').parentElement
    const styles = window.getComputedStyle(textInput as HTMLElement)

    expect(styles.marginTop).toBe('16px')
  })

  test('should trigger onChange handler', (): void => {
    const onChange = jest.fn()
    const { getByPlaceholderText } = render(
      <StyleProvider>
        <TextInput placeholder="get this" onChange={onChange} />
      </StyleProvider>
    )

    userEvent.type(getByPlaceholderText('get this'), 'typing...')
    expect(onChange).toHaveBeenCalled()
  })

  describe('with theme customization', (): void => {
    test('should have 2px border', (): void => {
      const theme = generateTheme({ primaryHue: 200, border: 'thick' })
      const { getByPlaceholderText } = render(
        <StyleProvider theme={theme}>
          <TextInput placeholder="Do I wanna know?" />
        </StyleProvider>
      )

      const textInput = getByPlaceholderText('Do I wanna know?')
      const styles = window.getComputedStyle(textInput)

      expect(styles.borderWidth).toBe('2px')
    })

    test('should have 8px border radius', (): void => {
      const theme = generateTheme({ primaryHue: 200, shape: 'intermediate' })
      const { getByPlaceholderText } = render(
        <StyleProvider theme={theme}>
          <TextInput placeholder="Do I wanna know?" />
        </StyleProvider>
      )

      const textInput = getByPlaceholderText('Do I wanna know?')
      const styles = window.getComputedStyle(textInput)

      expect(styles.borderRadius).toBe('8px')
    })

    test('should have 1px border radius', (): void => {
      const theme = generateTheme({ primaryHue: 200, shape: 'square' })
      const { getByPlaceholderText } = render(
        <StyleProvider theme={theme}>
          <TextInput placeholder="Do I wanna know?" />
        </StyleProvider>
      )

      const textInput = getByPlaceholderText('Do I wanna know?')
      const styles = window.getComputedStyle(textInput)

      expect(styles.borderRadius).toBe('1px')
    })
  })
})
