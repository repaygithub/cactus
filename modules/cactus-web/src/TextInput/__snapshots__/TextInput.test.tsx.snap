// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: TextInput should render a disabled input 1`] = `
<DocumentFragment>
  .c1 {
  border: 2px solid hsl(0,0%,90%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,90%);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
      disabled=""
    />
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should render a success input 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 20px;
  height: 20px;
  position: absolute;
  right: 8px;
  color: hsl(200,10%,20%);
}

.c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
  border-color: hsl(145,89%,28%);
  background: hsla(145,89%,28%,0.2);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
    />
    <svg
      class="c2"
      fill="currentcolor"
      height="1em"
      status="success"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should render a text input 1`] = `
<DocumentFragment>
  .c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
    />
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should render an error input 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 20px;
  height: 20px;
  position: absolute;
  right: 8px;
  color: hsl(200,10%,20%);
}

.c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
  border-color: hsl(353,84%,44%);
  background: hsla(353,84%,44%,0.2);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
    />
    <svg
      class="c2"
      fill="currentcolor"
      height="1em"
      status="error"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M5.82753204,18.84 L18.172832,18.84 L11.999732,6.4956 L5.82753204,18.84 Z M19.919732,21 L4.08063204,21 C3.70623204,21 3.35793204,20.8065 3.16083204,20.4888 C2.96463204,20.1702 2.94663204,19.7715 3.11403204,19.4376 L11.034032,3.5967 C11.216732,3.2313 11.591132,3 11.999732,3 C12.409232,3 12.783632,3.2313 12.966332,3.5967 L20.886332,19.4376 C21.053732,19.7715 21.034832,20.1702 20.839532,20.4888 C20.642432,20.8065 20.294132,21 19.919732,21 Z M11.9649,14.5902 C12.4419,14.5902 12.8298,14.322 12.8298,13.9917 L12.8298,10.7985 C12.8298,10.4682 12.4419,10.2 11.9649,10.2 C11.4879,10.2 11.1,10.4682 11.1,10.7985 L11.1,13.9917 C11.1,14.322 11.4879,14.5902 11.9649,14.5902 Z M12.9729,16.5369 C12.9729,16.0185 12.5544,15.6 12.036,15.6 C11.5185,15.6 11.1,16.0185 11.1,16.5369 C11.1,17.0544 11.5185,17.4738 12.036,17.4738 C12.5544,17.4738 12.9729,17.0544 12.9729,16.5369 Z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should render an input with a placeholder 1`] = `
<DocumentFragment>
  .c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
      placeholder="hold my place"
    />
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should render an warning input 1`] = `
<DocumentFragment>
  .c2 {
  vertical-align: middle;
  width: 24px;
  height: 24px;
  position: absolute;
  right: 8px;
  color: hsl(200,10%,20%);
}

.c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
  border-color: hsl(47,82%,47%);
  background: hsla(47,82%,47%,0.2);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
    />
    <svg
      class="c2"
      fill="currentcolor"
      height="1em"
      status="warning"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M11.9998875,21 C11.723584,21 11.4472806,20.8946987 11.2357779,20.683196 L3.31657896,12.763097 C2.89447368,12.3409918 2.89447368,11.6578832 3.31657896,11.2357779 L11.2357779,3.31657896 C11.6578832,2.89447368 12.3418918,2.89447368 12.763997,3.31657896 L20.683196,11.2357779 C21.1053013,11.6578832 21.1053013,12.3409918 20.683196,12.763097 L12.763997,20.683196 C12.5524944,20.8946987 12.276191,21 11.9998875,21 Z M5.60710759,11.9998875 L11.9998875,18.3926674 L18.3926674,11.9998875 L11.9998875,5.60620758 L5.60710759,11.9998875 Z M11.9641121,12.7903224 C11.4871061,12.7903224 11.1001013,12.522119 11.1001013,12.1909149 L11.1001013,8.99857498 C11.1001013,8.66737084 11.4871061,8.4000675 11.9641121,8.4000675 C12.441118,8.4000675 12.8281229,8.66737084 12.8281229,8.99857498 L12.8281229,12.1909149 C12.8281229,12.522119 12.441118,12.7903224 11.9641121,12.7903224 Z M12.9730247,14.7370467 C12.9730247,15.2554532 12.5545194,15.6739584 12.036113,15.6739584 C11.5186065,15.6739584 11.1001013,15.2554532 11.1001013,14.7370467 C11.1001013,14.2195402 11.5186065,13.800135 12.036113,13.800135 C12.5545194,13.800135 12.9730247,14.2195402 12.9730247,14.7370467 Z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`component: TextInput should support margin space props 1`] = `
<DocumentFragment>
  .c1 {
  border: 2px solid hsl(200,9%,35%);
  border-radius: 20px;
  height: 36px;
  outline: none;
  box-sizing: border-box;
  padding: 7px 28px 7px 15px;
  font-size: 18px;
  line-height: 1em;
  width: auto;
  background-color: hsl(0,0%,100%);
}

.c1:focus {
  border-color: hsl(200,96%,35%);
}

.c1::-webkit-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::-moz-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1:-ms-input-placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c1::placeholder {
  color: hsl(200,18%,45%);
  font-style: oblique;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 16px;
}

<div
    class="c0"
  >
    <input
      class="c0 c1"
    />
  </div>
</DocumentFragment>
`;
